docs(security): Add security documentation.

mprobst opened the pull request:
Substantially rewritten, based on original content by Brian Clarkio and
Ward Bell and contributions from Naomi Black.

naomiblack commented:
live preview at https://ngdocsdev.firebaseapp.com/docs/ts/latest/guide/security.html
Need some help with figuring out how to run and test the examples

Foxandxss commented:
I love this content.

To run e2e, you need to:

`$ gulp add-example-boilerplate` (to install the node_modules & typings on the example) and just:
`$ gulp e2e --filter=security`

For that to work there are some needed changes:

`e2e-spec.ts` needs a `'use strict';` because we are transpiling to ES6.

Then you need to run `$ gulp lint` because there are a couple of errors and in concrete one of them will make `tsc` not happy and not compile the example for the tests.

The examples have some errors too, I suggest running the example:

`$ cd public/docs/_examples/security/ts && npm start`

Once those errors are fixed, e2e should run fine.

Please ping me if you need something else 👍

mprobst commented:
@Foxandxss thanks, that helped a lot! Tests and code now fixed and working (fingers crossed).

@naomiblack I think this is good to go. I made some more typography fixes and included screenshots (and removed the obsolete ones).

mprobst commented:
@rjamet @koto @molnarg FYI. See live documentation at https://ngdocsdev.firebaseapp.com/docs/ts/latest/guide/security.html in particular.

Foxandxss commented:
Nice work. But there are still some lint errors.

Foxandxss commented:
Tests are fine, there is one failing but that is fixed in a different PR

mprobst commented:
Merging this now.

filipesilva commented on public/docs/ts/latest/guide/security.jade:
```diff
@@ -0,0 +1,246 @@
+  <a id="offline-template-compiler"></a>
+  ### Use the Offline Template Compiler
+
+  The offline template compiler prevents a whole class of vulnerabilities called template injection,
```
Do we have a resource available about how to use the offline template compiler?

mprobst replied:
Not yet. We should update once we have it.

filipesilva commented on public/docs/ts/latest/guide/security.jade:
```diff
@@ -0,0 +1,246 @@
+:marked
+  If we need to convert user input into a trusted value, it can be convenient to do so in a
+  controller method. The template below allows users to enter a YouTube video ID, and load the
+  corresponding video in an `<iframe>`. `<iframe src>` is a resource URL, because an untrusted
```
> `<iframe src>` is a resource URL, because an untrusted source can e.g. smuggle in file downloads that unsuspecting users would execute.

Weird phrase structure, not really sure of what it is trying to say.

mprobst replied:
Rephrased slightly.

filipesilva commented on public/docs/ts/latest/guide/security.jade:
```diff
@@ -0,0 +1,246 @@
+  cryptographically secure random number generator, and expire.
+
+  Angular applications can customize cookie and header names by binding their own
+  `CookieXSRFStrategy` value, or implement an entirely custom `XSRFStrategy` by providing a custom
```
How can users do this?

filipesilva commented:
Great guide!

mprobst commented:
Fixed.

naomiblack commented:
merged and live!

molnarg commented on public/docs/ts/latest/guide/security.jade:
```diff
@@ -0,0 +1,248 @@
+  US/docs/Web/Security/CSP/Introducing_Content_Security_Policy) is a defense-in-depth technique to
+  prevent XSS. To enable CSP, configure your web server to return an appropriate
+  `Content-Security-Policy` HTTP header. Learn more at
+  [OWASP](https://www.owasp.org/index.php/Content_Security_Policy).
```
I'm not sure if OWASP is the best source for an introduction to CSP. For example this is better: http://www.html5rocks.com/en/tutorials/security/content-security-policy/ @rjamet @koto do you know any good up to date CSP introduction article?

molnarg commented on public/docs/ts/latest/guide/security.jade:
```diff
@@ -0,0 +1,248 @@
+  Angular applications should follow the same security principles as regular web applications, and
+  should be audited as such. Angular specific APIs that should be audited in a security review,
+  such as the [_bypassSecurityTrust_](#bypass-security-apis) APIs, are marked in the documentation
+  as security sensitive.
```
Is there a way to list all APIs that are marked as security sensitive in the API documentation? It would be nice to include a link here if possible.

mprobst replied:
@naomiblack is working on that.

molnarg commented:
Overall I think it's a good guide and covers all important topic, thanks for writing it!

mprobst commented:
@molnarg addressing your comments in #1734.

koto commented:
Very good guide!
